File: qqzeng-ip-java/IpSearch.java
package qqzeng.ip;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
/**
 * 最新一代文件结构 高性能解析IP数据库 qqzeng-ip.dat
 * 编码：UTF8 字节序：Little-Endian
 * For detailed information and guide: http://qqzeng.com/
 * @author qqzeng-ip 于 2015-08-01
 */
public class IpSearch {

	private static IpSearch instance = null;

	private byte[] data;

	private HashMap<Integer, PrefixIndex> prefixMap;

	private long firstStartIpOffset;// 索引区第一条流位置
	// private int lastStartIpOffset;//索引区最后一条流位置
	private long prefixStartOffset;// 前缀区第一条的流位置
	private long prefixEndOffset;// 前缀区最后一条的流位置
	// private int ipCount; //ip段数量
	private long prefixCount; // 前缀数量

	public IpSearch() {

		Path path = Paths
				.get("C:/qqzeng-ip-java/bin/qqzeng-ip.dat");

		try {
			data = Files.readAllBytes(path);
		} catch (IOException e) {
			e.printStackTrace();
		}

		firstStartIpOffset = BytesToLong(data[0], data[1], data[2], data[3]);
		// lastStartIpOffset = BytesToLong(data[4], data[5], data[6], data[7]);
		prefixStartOffset = BytesToLong(data[8], data[9], data[10], data[11]);
		prefixEndOffset = BytesToLong(data[12], data[13], data[14], data[15]);

		// ipCount = (lastStartIpOffset - firstStartIpOffset) / 12 + 1; //索引区块每组
		// 12字节
		prefixCount = (prefixEndOffset - prefixStartOffset) / 9 + 1; // 前缀区块每组
		// 9字节

		// 初始化前缀对应索引区区间
		byte[] indexBuffer = Arrays.copyOfRange(data, (int) prefixStartOffset,(int) prefixEndOffset + 9);
		prefixMap = new HashMap<Integer, PrefixIndex>();
		for (int k = 0; k < prefixCount; k++) {
			int i = k * 9;
			int prefix = (int) (indexBuffer[i] & 0xFFL);

			PrefixIndex pf = new PrefixIndex();
			pf.start_index = BytesToLong(indexBuffer[i + 1],indexBuffer[i + 2], indexBuffer[i + 3], indexBuffer[i + 4]);
			pf.end_index = BytesToLong(indexBuffer[i + 5], indexBuffer[i + 6],indexBuffer[i + 7], indexBuffer[i + 8]);
			prefixMap.put(prefix, pf);

		}

	}

	public synchronized static IpSearch getInstance() {
		if (null == instance)
			instance = new IpSearch();
		return instance;
	}

	public String Get(String ip) {
		String[] ips = ip.split("\\.");
		int prefix = Integer.valueOf(ips[0]);
		long intIP = ipToLong(ip);

		long high = 0;
		long low = 0;

		if (prefixMap.containsKey(prefix)) {
			low = prefixMap.get(prefix).start_index;
			high = prefixMap.get(prefix).end_index;

		} else {
			return "";
		}
		
		long my_index = low == high ? low : BinarySearch(low, high, intIP);

		IpIndex ipindex = new IpIndex();
		GetIndex((int) my_index, ipindex);

		if ((ipindex.startip <= intIP) && (ipindex.endip >= intIP)) {
			return GetLocal(ipindex.local_offset, ipindex.local_length);
		} else {
			return "";
		}

	}

	// / <summary>
	// / 二分逼近算法
	// / </summary>
	public long BinarySearch(long low, long high, long k) {
		long M = 0;
		while (low <= high) {
			long mid = (low + high) / 2;

			long endipNum = GetEndIp(mid);
			if (endipNum >= k) {
				M = mid;
				if (mid == 0) {
					break; // 防止溢出
				}
				high = mid - 1;
			} else
				low = mid + 1;
		}
		return M;
	}

	// / <summary>
	// / 在索引区解析
	// / </summary>
	// / <param name="left">ip第left个索引</param>
	private void GetIndex(int left, IpIndex ipindex) {
		int left_offset = (int) firstStartIpOffset + (left * 12);
		ipindex.startip = BytesToLong(data[left_offset], data[1 + left_offset],data[2 + left_offset], data[3 + left_offset]);
		ipindex.endip = BytesToLong(data[4 + left_offset],data[5 + left_offset], data[6 + left_offset],data[7 + left_offset]);
		ipindex.local_offset = (int) BytesToLong3(data[8 + left_offset],data[9 + left_offset], data[10 + left_offset]);
		ipindex.local_length = (int) data[11 + left_offset];
	}

	// / <summary>
	// / 只获取结束ip的数值
	// / </summary>
	// / <param name="left">索引区第left个索引</param>
	// / <returns>返回结束ip的数值</returns>
	private long GetEndIp(long left) {
		int left_offset = (int) firstStartIpOffset + (int) (left * 12);
		return BytesToLong(data[4 + left_offset], data[5 + left_offset],data[6 + left_offset], data[7 + left_offset]);

	}

	// / <summary>
	// / 返回地址信息
	// / </summary>
	// / <param name="local_offset">地址信息的流位置</param>
	// / <param name="local_length">地址信息的流长度</param>
	// / <returns></returns>
	private String GetLocal(int local_offset, int local_length) {
		byte[] bytes = new byte[local_length];
		bytes = Arrays.copyOfRange(data, local_offset, local_offset	+ local_length);
		return new String(bytes, StandardCharsets.UTF_8);

	}

	// / <summary>
	// / 字节转整形 小节序
	// / </summary>
	private long BytesToLong(byte a, byte b, byte c, byte d) {
		return (a & 0xFFL) | ((b << 8) & 0xFF00L) | ((c << 16) & 0xFF0000L)	| ((d << 24) & 0xFF000000L);

	}

	private long BytesToLong3(byte a, byte b, byte c) {
		return (a & 0xFFL) | ((b << 8) & 0xFF00L) | ((c << 16) & 0xFF0000L);

	}

	public long ipToLong(String ip) {
		String[] quads = ip.split("\\.");
		long result = 0;
		result += Integer.parseInt(quads[3]);
		result += Long.parseLong(quads[2]) << 8L;
		result += Long.parseLong(quads[1]) << 16L;
		result += Long.parseLong(quads[0]) << 24L;
		return result;
	}

	public static void main(String[] args) {

		IpSearch finder = IpSearch.getInstance();

		String ip = "210.51.200.123";
		String result = finder.Get(ip);
		System.out.println(ip);
		System.out.println(result);

		/*
		 * 1.197.224.9 亚洲|中国|河南|周口|商水|电信|411623|China|CN|114.60604|33.53912
		 */

	}
}

File: qqzeng-ip-2.0/qqzeng-ip-go/ipsearch.go
package ipsearch

import (
	"io/ioutil"
	"log"
	"strconv"
	"strings"
)

/**
 * @author xiao.luo
 * @description This is the go version for IpSearch
 */

type ipIndex struct {
	startip, endip             uint32
	local_offset, local_length uint32
}

type prefixIndex struct {
	start_index, end_index uint32
}

type ipSearch struct {
	data               []byte
	prefixMap          map[uint32]prefixIndex
	firstStartIpOffset uint32
	prefixStartOffset  uint32
	prefixEndOffset    uint32
	prefixCount        uint32
}

var ips *ipSearch = nil

func New() (ipSearch, error) {
	if ips == nil {
		var err error
		ips, err = loadIpDat()
		if err != nil {
			log.Fatal("the IP Dat loaded failed!")
			return *ips, err
		}
	}
	return *ips, nil
}

func loadIpDat() (*ipSearch, error) {

	p := ipSearch{}
	//加载ip地址库信息
	data, err := ioutil.ReadFile("F:/ClearClouds/Projects/Java/kepi/src/main/resources/regionIp.dat")
	if err != nil {
		log.Fatal(err)
	}
	p.data = data
	p.prefixMap = make(map[uint32]prefixIndex)

	p.firstStartIpOffset = bytesToLong(data[0], data[1], data[2], data[3])
	p.prefixStartOffset = bytesToLong(data[8], data[9], data[10], data[11])
	p.prefixEndOffset = bytesToLong(data[12], data[13], data[14], data[15])
	p.prefixCount = (p.prefixEndOffset-p.prefixStartOffset)/9 + 1 // 前缀区块每组

	// 初始化前缀对应索引区区间
	indexBuffer := p.data[p.prefixStartOffset:(p.prefixEndOffset + 9)]
	for k := uint32(0); k < p.prefixCount; k++ {
		i := k * 9
		prefix := uint32(indexBuffer[i] & 0xFF)

		pf := prefixIndex{}
		pf.start_index = bytesToLong(indexBuffer[i+1], indexBuffer[i+2], indexBuffer[i+3], indexBuffer[i+4])
		pf.end_index = bytesToLong(indexBuffer[i+5], indexBuffer[i+6], indexBuffer[i+7], indexBuffer[i+8])
		p.prefixMap[prefix] = pf

	}
	return &p, nil
}

func (p ipSearch) Get(ip string) string {
	ips := strings.Split(ip, ".")
	x, _ := strconv.Atoi(ips[0])
	prefix := uint32(x)
	intIP := ipToLong(ip)

	var high uint32 = 0
	var low uint32 = 0

	if _, ok := p.prefixMap[prefix]; ok {
		low = p.prefixMap[prefix].start_index
		high = p.prefixMap[prefix].end_index
	} else {
		return ""
	}

	var my_index uint32
	if low == high {
		my_index = low
	} else {
		my_index = p.binarySearch(low, high, intIP)
	}

	ipindex := ipIndex{}
	ipindex.getIndex(my_index, &p)

	if ipindex.startip <= intIP && ipindex.endip >= intIP {
		return ipindex.getLocal(&p)
	} else {
		return ""
	}
}

// 二分逼近算法
func (p ipSearch) binarySearch(low uint32, high uint32, k uint32) uint32 {
	var M uint32 = 0
	for low <= high {
		mid := (low + high) / 2

		endipNum := p.getEndIp(mid)
		if endipNum >= k {
			M = mid
			if mid == 0 {
				break // 防止溢出
			}
			high = mid - 1
		} else {
			low = mid + 1
		}
	}
	return M
}

// 只获取结束ip的数值
// 索引区第left个索引
// 返回结束ip的数值
func (p ipSearch) getEndIp(left uint32) uint32 {
	left_offset := p.firstStartIpOffset + left*12
	return bytesToLong(p.data[4+left_offset], p.data[5+left_offset], p.data[6+left_offset], p.data[7+left_offset])

}

func (p *ipIndex) getIndex(left uint32, ips *ipSearch) {
	left_offset := ips.firstStartIpOffset + left*12
	p.startip = bytesToLong(ips.data[left_offset], ips.data[1+left_offset], ips.data[2+left_offset], ips.data[3+left_offset])
	p.endip = bytesToLong(ips.data[4+left_offset], ips.data[5+left_offset], ips.data[6+left_offset], ips.data[7+left_offset])
	p.local_offset = bytesToLong3(ips.data[8+left_offset], ips.data[9+left_offset], ips.data[10+left_offset])
	p.local_length = uint32(ips.data[11+left_offset])
}

// / 返回地址信息
// / 地址信息的流位置
// / 地址信息的流长度
func (p *ipIndex) getLocal(ips *ipSearch) string {
	bytes := ips.data[p.local_offset : p.local_offset+p.local_length]
	return string(bytes)

}

func ipToLong(ip string) uint32 {
	quads := strings.Split(ip, ".")
	var result uint32 = 0
	a, _ := strconv.Atoi(quads[3])
	result += uint32(a)
	b, _ := strconv.Atoi(quads[2])
	result += uint32(b) << 8
	c, _ := strconv.Atoi(quads[1])
	result += uint32(c) << 16
	d, _ := strconv.Atoi(quads[0])
	result += uint32(d) << 24
	return result
}

//字节转整形
func bytesToLong(a, b, c, d byte) uint32 {
	a1 := uint32(a)
	b1 := uint32(b)
	c1 := uint32(c)
	d1 := uint32(d)
	return (a1 & 0xFF) | ((b1 << 8) & 0xFF00) | ((c1 << 16) & 0xFF0000) | ((d1 << 24) & 0xFF000000)
}

func bytesToLong3(a, b, c byte) uint32 {
	a1 := uint32(a)
	b1 := uint32(b)
	c1 := uint32(c)
	return (a1 & 0xFF) | ((b1 << 8) & 0xFF00) | ((c1 << 16) & 0xFF0000)

}
